### PR TITLE
CL-643: Generalize more DAO business logic by introducing low level database API in DAOs

### PR DESCRIPTION
**Pull request by plyhun**
## Abstract

In order to put more business logic off the DB-dependent impls into an API, we need to introduce some lower-level database API, for the raw persistent entity management. Otherwise the BL has to be copied around for every DB implementation.

Changes:
1. Some type inconsistency is fixed (`HibNamedElement` is an independent contract, as initially)
2. The references to `HibCoreElement` are told apart with `HibNamedElement`, where applicable.
3. ~~NEW: `DaoPersistable` containing low level access API to the context independent entities. `OrientDBDaoGlobal`/`AbstractCoreDaoWrapper` partially implement the new API.~~ Change requested. See point 7
4. ~~NEW `RootDaoPersistable` containing low level access API to the context-dependent (mainly:project-dependent) entities. `OrientDBRootDao`/`AbstractRootDaoWrapper` partially implement the new API.~~ Change requested. See point 7
5. The existing entity creation code is updated with the new API access (only `(User|Schema|Microschema)DaoWrapperImpl` by now).
6. The updated DAO code is moved into MDM API (only `(User|Schema|Microschema)Dao` by now).
7. NEW TxEntityPersistenceManager contains low level persistence API for context-independent/dependent/agnostic entities. This has been moved out of DAO API to tell apart the high level (DAO) and low level (database) entity management.
8. As a part of DAO API, the `FieldSchemaChange` functionality has been generalized. This includes `Hib*` interfaces, BL moved into the API, some deprecated/duplicated code removal (new DAO implementations vs old Entity implementations).

### General

* [x] Added abstract that describes the change
* [x] Added changelog entry to `/CHANGELOG.adoc`
* [x] Ensured that the change is covered by tests
* [x] Ensured that the change is documented in the docs

### On API Changes

* [x] Checked if the changes are breaking or not
* [x] Added GraphQL API if applicable
* [x] Added Elasticsearch mapping if applicable

### CODE REVIEW

File: mdm/common/src/main/java/com/gentics/mesh/dagger/MeshBuilderFactory.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.gentics.mesh.dagger;
```
Comment (npomaroli) [inline mdm/common/src/main/java/com/gentics/mesh/dagger/MeshBuilderFactory.java:1]: This is a duplicate of MeshBuilderFactory in core, should this have been moved here?

File: mdm/common/src/main/java/com/gentics/mesh/dagger/MeshComponent.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package com.gentics.mesh.dagger;
```
Comment (npomaroli) [inline mdm/common/src/main/java/com/gentics/mesh/dagger/MeshComponent.java:1]: This is a duplicate of MeshComponent in core, should this have been moved here?

File: mdm/common/src/main/java/com/gentics/mesh/dagger/SearchProviderType.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.gentics.mesh.dagger;
```
Comment (npomaroli) [inline mdm/common/src/main/java/com/gentics/mesh/dagger/SearchProviderType.java:1]: This is a duplicate of SearchProviderType in core.

File: mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractContainerDaoWrapper.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import com.gentics.mesh.core.data.schema.HibSchemaChange;
+import com.gentics.mesh.core.rest.common.NameUuidReference;
+import com.gentics.mesh.core.rest.schema.FieldSchemaContainer;
+import com.gentics.mesh.core.rest.schema.FieldSchemaContainerVersion;
+import com.gentics.mesh.core.rest.schema.change.impl.SchemaChangeModel;
+import com.gentics.mesh.core.result.Result;
+
+import dagger.Lazy;
+
+public abstract class AbstractContainerDaoWrapper<
```
Comment (npomaroli) [inline mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractContainerDaoWrapper.java:21]: Javadoc

File: mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractCoreDaoWrapper.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * @author plyhun
+ *
+ * @param <R> REST entity model
+ * @param <T> MDM API entity counterpart
+ * @param <D> OrientDB Vertex entity counterpart
+ */
+public abstract class AbstractCoreDaoWrapper<R extends RestModel, T extends HibCoreElement<R>, D extends MeshCoreVertex<R>> 
+		extends AbstractDaoWrapper<T> implements DaoGlobal<T>, DaoTransformable<T, R> {
+
+	public AbstractCoreDaoWrapper(Lazy<OrientDBBootstrapInitializer> boot, Lazy<PermissionPropertiesImpl> permissions) {
```
Comment (npomaroli) [inline mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractCoreDaoWrapper.java:26]: Javadoc

File: mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractRootDaoWrapper.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * @param <RM> REST leaf entity model
+ * @param <L> MDM API leaf entity counterpart
+ * @param <D> OrientDB Vertex leaf entity counterpart
+ * @param <R> MDM API root entity type 
+ */
+public abstract class AbstractRootDaoWrapper<RM extends RestModel, L extends HibCoreElement<RM>, D extends MeshCoreVertex<RM>, R extends HibCoreElement<? extends RestModel>> 
+	extends AbstractDaoWrapper<L> implements DaoTransformable<L, RM> {
+
+	public AbstractRootDaoWrapper(Lazy<OrientDBBootstrapInitializer> boot, Lazy<PermissionPropertiesImpl> permissions) {
```
Comment (npomaroli) [inline mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/data/dao/AbstractRootDaoWrapper.java:27]: Javadoc

File: mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/db/AbstractTx.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
+/**
```
Comment (npomaroli) [inline mdm/orientdb-wrapper/src/main/java/com/gentics/mesh/core/db/AbstractTx.java:1]: This is a duplicate of AbstractTx in mesh-common

**Review by npomaroli — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/gentics/mesh/core/db/AbstractTx.java [hunk 8]
```diff
@@ -108,4 +119,71 @@ protected void setGraph(T currentGraph) {
...
+		/*
+		 * Since OrientDB does not tell apart POJOs and persistent entities, 
+		 * processing the entity updates directly into the persistent state, 
+		 * the merge implementation here is empty.
+		 */
+		return element;
+	}
+	
+	@Override
+	public <B extends HibBaseElement> void delete(B element, Class<? extends B> classOfB) {
```
[the remark this answers is not in the document]
Comment (plyhun) [reply common/src/main/java/com/gentics/mesh/core/db/AbstractTx.java:186]: -> https://github.com/gentics/mesh/pull/1262

File: common/src/main/java/com/gentics/mesh/core/db/AbstractTx.java [hunk 9]
```diff
@@ -108,4 +119,71 @@ protected void setGraph(T currentGraph) {
 		this.currentGraph = currentGraph;
 	}
 
+	@SuppressWarnings("unchecked")
+	@Override
+	public <B extends HibCoreElement<? extends RestModel>> B create(String uuid, Dao<B> dao) {
+		return ((AbstractCoreDaoWrapper<?, ? extends B, ? extends B>) dao).persist(uuid);
```
[the remark this answers is not in the document]
Comment (plyhun) [reply common/src/main/java/com/gentics/mesh/core/db/AbstractTx.java:125]: -> https://github.com/gentics/mesh/pull/1262

**Pull request merged**